### ISSUE TITLE
IngressRoute does not honor kubernetes.io/ingress.class

### ISSUE THREAD

**Issue by pimvanpelt** [heptio/contour#720]
**What steps did you take and what happened:**
Running two contour pools, one called `contour-prod` and the other `contour-staging` (each setting the appropriate flag in `--ingress-class-name`

Two problems that signal that the annotation is not honored:

### Problem 1

Setting `annotations.kubernetes.io/ingress.class: contour-staging` on an IngressRoute applies this to all running contour instances, not the specific one selected.

```
$ kubectl -n contour get svc
NAME              TYPE           CLUSTER-IP     EXTERNAL-IP      PORT(S)                      AGE
contour-prod      LoadBalancer   10.109.26.45   194.126.235.48   80:30113/TCP,443:32570/TCP   3h42m
contour-staging   LoadBalancer   10.108.80.15   192.168.8.128    80:31054/TCP,443:30366/TCP   3h45m

$ kubectl get ingressroute --all-namespaces
NAMESPACE   NAME            FQDN          TLS SECRET   FIRST ROUTE   STATUS   STATUS DESCRIPTION
nginx       nginx-staging   k8s.ipng.nl                /             valid    valid IngressRoute
```

This IngressRoute is as expected, created on nginx-staging:
```
$ curl -I -H "Host: k8s.ipng.nl" 192.168.8.128
HTTP/1.1 200 OK
server: envoy
date: Tue, 02 Oct 2018 22:23:09 GMT
content-type: text/html
content-length: 612
last-modified: Tue, 03 Jul 2018 13:27:08 GMT
etag: "5b3b79ac-264"
accept-ranges: bytes
x-envoy-upstream-service-time: 5
```

This IngressRoute is NOT expected, it is not annotated to be created on nginx-prod:
```
$ curl -I -H "Host: k8s.ipng.nl" 194.126.235.48
HTTP/1.1 200 OK
server: envoy
date: Tue, 02 Oct 2018 22:24:15 GMT
content-type: text/html
content-length: 612
last-modified: Tue, 03 Jul 2018 13:27:08 GMT
etag: "5b3b79ac-264"
accept-ranges: bytes
x-envoy-upstream-service-time: 4
```

### Problem 2
Creating two IngressRoutes, one with contour-staging and one with contour-prod annotated, yields a conflict because both picked them up.

```
$ kubectl get ingressroute --all-namespaces
NAMESPACE   NAME            FQDN          TLS SECRET   FIRST ROUTE   STATUS    STATUS DESCRIPTION
nginx       nginx-prod      k8s.ipng.nl                /             invalid   fqdn "k8s.ipng.nl" is used in multiple IngressRoutes: nginx/nginx-prod, nginx/nginx-staging
nginx       nginx-staging   k8s.ipng.nl                /             invalid   fqdn "k8s.ipng.nl" is used in multiple IngressRoutes: nginx/nginx-prod, nginx/nginx-staging
```

**What did you expect to happen:**

I expect in problem 1, that the curl on the contour-prod service fails.
I expect in problem 2, that:
*    the curl on contour-prod service yields a response from nginx-prod, and that the curl on contour-staging service yields a response from nginx-staging.
*    the creation of two IngressRoutes (one for nginx-prod -> contour-prod and one for nginx-staging -> contour-staging)

**Anything else you would like to add:**

This problem is specific to the IngressRoute CRD. Normal Ingress objects do honor the annotation as per documentation, and I can successfully tie the nginx-prod -> contour-prod and nginx-staging -> contour-staging as per my expectations.

**Environment:**

- Contour version: 
docker-pullable://gcr.io/heptio-images/contour@sha256:8ef6d4f97ad678fab75fde095ee924b83881ade5acef37c80e485a02f12e9a37

- Kubernetes version: (use `kubectl version`):
Client Version: version.Info{Major:"1", Minor:"12", GitVersion:"v1.12.0", GitCommit:"0ed33881dc4355495f623c6f22e7dd0b7632b7c0", GitTreeState:"clean", BuildDate:"2018-09-27T17:05:32Z", GoVersion:"go1.10.4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"12", GitVersion:"v1.12.0", GitCommit:"0ed33881dc4355495f623c6f22e7dd0b7632b7c0", GitTreeState:"clean", BuildDate:"2018-09-27T16:55:41Z", GoVersion:"go1.10.4", Compiler:"gc", Platform:"linux/amd64"}

- Kubernetes installer & version:
N/A.

- Cloud provider or hardware configuration:
Baremetal; MetalLB loadbalancer.

- OS (e.g. from `/etc/os-release`):
Ubuntu Bionic LTS.

**Comment by amoskyler** [comment]
I can confirm this issue. I was able to replicate it.

**Comment by davecheney** [comment]
Thanks for raising this issue. 

Ingressroute crds aren’t subject to the ingress class annotation; they aren’t ingress objects. 

Rather than slap on support for an ingressroute version of ingress class, could you please explain the problem you wanted to solve by reaching for ingress class? That helps me understand the underlying requirement.

**Comment by amoskyler** [comment]
Ah, the distinction between ingressroute and ingress annotations did not occur to me - makes sense.

My use case:
I use ingress class to isolate traffic for different edge load balancers. I use both an internal vpc load balancer and an external load balancer for ingressing public internet traffic. Each ingress is responsible for the routing of it's corresponding ELB. 
Without the ability to distinguish which controller should handle specific routes, my public facing ELB/ingress will serve traffic to internal services. There may be a better way to handle this, but I haven't been able to come up with anything particularly viable other than running two clusters.

I currently solve for this using the Ingress resource with the ingress-class annotation.

Ideally only routes which are assigned an ingress class will only be included in the DAG of that specific controller, and ignored by all others.

**Comment by stevesloka** [comment]
@pimvanpelt so you're using the same cluster for all environments? 

One way you could solve this without needing 3 controllers running is to utilize delegation. Configure Contour to only allow Root ingress routes in a specific namespace (this piece is optional but safer since it has more restrictions on who can create `root` IngressRoutes). Have your VHosts (root IngressRoutes) configured there for each environment and delegate them to your matching namespaces per environment. 

For example:
- dev.com --> Namespace `dev`
- stage.com --> Namespace `stage`
- prod.com --> Namespace `prod`

By doing this, any IngressRoute created within each environment can only utilize paths off of the Vhost that is delegated to them. 

The only downside which may be a deal breaker for you would be that you would mix the traffic for all the environments by implementing this way.

**Comment by pimvanpelt** [comment]
@stevesloka Yes, we are using the same cluster.

The traffic isolation is a mild concern, which we don't have today because we're using three Ingress pools. I haven't really gotten enough experience with IngressRoute, so I'll play around with your proposal. Envoy is not the performance bottleneck for us, so my concern on traffic isolation is a nice-to-have.

**Comment by davecheney** [comment]
Here are a few thing that I think are possible for 0.7 which don't paint us into a corner in the future.

1. Contour already supports the `kubernetes.io/ingress.class` annotation, but this is intended for running Contour in conjunction with other ingress controllers, so you can trial contour on your current cluster without betting the farm.
2. `contour serve` supports a  `--ingress-class-name=INGRESS-CLASS-NAME` flag which is intended to be used to allow contour to impersonate another ingress controller. I added this mainly to work around kube-lego's hard coded list of known ingress controllers.
3. Using an annotation on IngressRoute resources is probably the way to go, this does feel like an annotation sort of meta data type configuration parameter.
4. Whatever we choose we need to be mindful of the `--ingressroute-root-namespaces=INGRESSROUTE-ROOT-NAMESPACES` flag, I want to avoid confusing, thus hard to document and harder to internalise, interactions between those flags.

Here is what I propose

a. We add a new annotation `contour.heptio.com/ingress.class` which is recognised on both `ingress` and `ingressroute` objects. #739 does this.
b. We repurpose `--ingress-class-name=INGRESS-CLASS-NAME` to also change the name that the `contour.heptio.com/ingress.class` annotation matches on. #739 does this.
c. At some point in the future the `--ingress-class-name=INGRESS-CLASS-NAME` flag should move to a config map. This does not need to happen for 0.7

**Comment by stevesloka** [comment]
Could we have a label selector on namespaces, so each instance of contour would only select resources in namespaces that match a label selector?

**Comment by davecheney** [comment]
That's a good idea, but I'm concerned with the interaction between namespaces and labels. If we say "this contour looks for label `quux` across all namespaces in the cluster" then that's basically an ingress.class annotation in a cheap suit. If we let people nominate a namespace:label combination, then where do we stick that data, on the namespace object?

**Comment by stevesloka** [comment]
Yup would work like services do, labels on a namespace to determine the ingress class. contour then just looks at labels when a resource is discovered. New namespaces will be added to the ingress class automatically. This then keeps the Ingressroute objects generic and non environment specific.

Just thinking out loud looking for a way to not require all the annotations.

**Comment by mauilion** [comment]
Another couple use cases related to #410 

When we have "default tls" wired in it would be helpful to consider it's usage in this case as well.

I know of a few folks that use a wildcard pattern and delegate the host side of the fqdn to teams. For example:

Attract `*.app.com` to the ingress controller presumably with a wildcard dns record. Set `--ingress-class=app-com` on the ingress controller and provide a default cert that has *.app.com as a san.

Then inform the team that when creating a new ingress resource they can follow that fqdn pattern. 
Make a new ingress resource like:
```
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  annotations:
    kubernetes.io/ingress.class: app-com
  name: web
spec:
  tls:
  - hosts:
    - team1.app.com
  rules:
    - host: team1.app.com
      http:
        paths:
          - backend:
              serviceName: web
              servicePort: 80
            path: /
```
This enables tls for the ingress resource and since the secret is not defined the controller uses the "default" cert for this and secures it with *.app.com 

I may want to service other domains this way as well.

I like @stevesloka idea about label match on namespaces but this pattern would break that. As in most cases I would want to "call out" which ingress controller is wired right to satisfy this request.

I can't wait to see what schemes we can cook up to solve this with ingressroute/contour :) 
Y'all are awesome!

**Comment by davecheney** [comment]
Closed in #739
